@here/harp-materials/lib/MapMeshMaterials.ts

```diff
@@ -706,16 +857,34 @@ export class ExtrusionFeatureMixin implements ExtrusionFeature {
      * Apply the extrusionRatio value from the parameters to the respective properties.
      */
     protected applyExtrusionParameters(params?: ExtrusionFeatureParameters) {
+        // Prepare maps for holding uniforms and defines references from the actual material.
+        linkMixinWithMaterial(this, this);
+
+        assert(this.shaderDefines !== undefined);
+        assert(this.shaderUniforms !== undefined);
+
+        // Create uniform with default value, this ensures that it is always created,
+        // so no need for checks in setters.
+        this.shaderUniforms!.extrusionRatio = new THREE.Uniform(
+            ExtrusionFeatureDefs.DEFAULT_RATIO_MAX
+        );
+
         // Apply initial parameter values.
         if (params !== undefined) {
             if (params.extrusionRatio !== undefined) {
                 this.setExtrusionRatio(params.extrusionRatio);
             }
         }
 
-        (this as any).onBeforeCompile = (shader: any) => {
-            ExtrusionFeature.onBeforeCompile(this, shader);
-        };
+        const material = this as any;
+        material.onBeforeCompile = chainCallbacks(
+            material.onBeforeCompile,
+            (shader: THREE.Shader) => {
+                ExtrusionFeature.onBeforeCompile(this, shader);
+            }
+        );
+
+        this.needsUpdate = ExtrusionFeature.isEnabled(this);
```

Comment:
Is this call really necessary?  Shouldn't `needUpdate` already be set in `ExtrusionFeature.updateExtrusionFeature()`?

Reply:
It is set in namespace feature update function:
`ExtrusionFeature.updateExtrusionFeature()`,
but here we have flow initiated from the MapMeshMaterial c-tor, and there is no warranty that updateExtrusionFeature will be called at all, if no c-tor `params` are defined the corresponding setter is not called:
`this.setExtrusionRatio()`, thus update will not be performed, which is fine.
The c-tor will re-build material only if the feature is enabled, if not we create uniforms, prepare shader references and wait for values update (via setters) that in turn may activate (material re-build) features. Otherwise the original material is preserved, with handy wrapper on it (mixin feature).